File: internal/app/selection/selection.go
package selection

import (
	"sort"

	"github.com/freesiapro/resize-to-telegram-sticker/internal/app/job"
	"github.com/freesiapro/resize-to-telegram-sticker/internal/domain"
	"github.com/freesiapro/resize-to-telegram-sticker/internal/infra"
)

type SelectionItem struct {
	Path  string
	IsDir bool
}

type ExpandResult struct {
	Jobs       []job.Job
	DirCount   int
	FileCount  int
	TotalFiles int
	OutputDirs []string
	Skipped    []job.Skipped
}

type SelectionExpander struct {
	ListFiles func(root string) ([]string, error)
}

func (e SelectionExpander) Expand(selections []SelectionItem, outputDir string) (ExpandResult, error) {
	result := ExpandResult{}
	if outputDir == "" {
		outputDir = "./output"
	}
	listFiles := e.ListFiles
	if listFiles == nil {
		listFiles = infra.ListFiles
	}

	jobs := make([]job.Job, 0)
	seen := make(map[string]struct{})
	outputSet := make(map[string]struct{})

	files := make([]SelectionItem, 0)
	dirs := make([]SelectionItem, 0)
	for _, s := range selections {
		if s.IsDir {
			dirs = append(dirs, s)
			continue
		}
		files = append(files, s)
	}

	for _, s := range files {
		kind, err := domain.DetectInputKind(s.Path)
		if err != nil {
			result.Skipped = append(result.Skipped, job.Skipped{Path: s.Path, Reason: err.Error()})
			continue
		}
		if _, ok := seen[s.Path]; ok {
			continue
		}
		seen[s.Path] = struct{}{}
		jobs = append(jobs, job.Job{InputPath: s.Path, Kind: kind, OutputDir: outputDir})
		result.FileCount++
		result.TotalFiles++
		outputSet[outputDir] = struct{}{}
	}

	for _, s := range dirs {
		filesInDir, err := listFiles(s.Path)
		if err != nil {
			return ExpandResult{}, err
		}
		result.DirCount++
		for _, path := range filesInDir {
			kind, err := domain.DetectInputKind(path)
			if err != nil {
				result.Skipped = append(result.Skipped, job.Skipped{Path: path, Reason: err.Error()})
				continue
			}
			if _, ok := seen[path]; ok {
				continue
			}
			seen[path] = struct{}{}
			jobs = append(jobs, job.Job{InputPath: path, Kind: kind, OutputDir: outputDir})
			result.TotalFiles++
			outputSet[outputDir] = struct{}{}
		}
	}

	result.Jobs = jobs
	result.OutputDirs = sortedKeys(outputSet)
	return result, nil
}

func sortedKeys(set map[string]struct{}) []string {
	keys := make([]string, 0, len(set))
	for k := range set {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: python/core/selection.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable

from core.job import Job, Skipped
from core.media import detect_input_kind
from infra.files import list_files


@dataclass(frozen=True)
class SelectionItem:
    path: str
    is_dir: bool


@dataclass
class ExpandResult:
    jobs: list[Job] = field(default_factory=list)
    dir_count: int = 0
    file_count: int = 0
    total_files: int = 0
    output_dirs: list[str] = field(default_factory=list)
    skipped: list[Skipped] = field(default_factory=list)


class SelectionExpander:
    def __init__(self, list_files_fn: Callable[[str], list[str]] | None = None) -> None:
        self._list_files = list_files_fn or list_files

    def expand(self, selections: list[SelectionItem], output_dir: str) -> ExpandResult:
        result = ExpandResult()
        if not output_dir:
            output_dir = "./output"

        jobs: list[Job] = []
        seen: set[str] = set()
        output_set: set[str] = set()

        files: list[SelectionItem] = []
        dirs: list[SelectionItem] = []
        for selection in selections:
            if selection.is_dir:
                dirs.append(selection)
            else:
                files.append(selection)

        for selection in files:
            try:
                kind = detect_input_kind(selection.path)
            except ValueError as exc:
                result.skipped.append(Skipped(path=selection.path, reason=str(exc)))
                continue
            if selection.path in seen:
                continue
            seen.add(selection.path)
            jobs.append(Job(input_path=selection.path, kind=kind, output_dir=output_dir))
            result.file_count += 1
            result.total_files += 1
            output_set.add(output_dir)

        for selection in dirs:
            files_in_dir = self._list_files(selection.path)
            result.dir_count += 1
            for path in files_in_dir:
                try:
                    kind = detect_input_kind(path)
                except ValueError as exc:
                    result.skipped.append(Skipped(path=path, reason=str(exc)))
                    continue
                if path in seen:
                    continue
                seen.add(path)
                jobs.append(Job(input_path=path, kind=kind, output_dir=output_dir))
                result.total_files += 1
                output_set.add(output_dir)

        result.jobs = jobs
        result.output_dirs = sorted(output_set)
        return result
